### ISSUE TITLE
Docker containers can become zombies

### ISSUE THREAD

**Issue by fnothaft** [BD2KGenomics/toil#987]
If a job terminates but that job was running docker containers, the docker containers can persist in the background and not get killed off.

Depends on #1189 (PR #1199)

**Comment by fnothaft** [comment]
As suggested by @hannes-ucsc, look at https://docs.docker.com/swarm/ before doing anything.

**Comment by hannes-ucsc** [comment]
@benedictpaten to write up his proposal here.

**Comment by hannes-ucsc** [comment]
Renaming to zombie as that is established Unix terminology for processes and Docker containers are processes.

**Comment by arkal** [comment]
@benedictpaten and I had a long chat yesterday and we built on the chat @hannes-ucsc and I had last week. We will use the inherent caching mechanism to track pids of jobs and then use the filestore context manager to allow jobs to clean up behind themselves on completion in a finally clause. On entering the filestore, jobs will check for other jobs' integrity and purge any dead jobs. Users will have the ability to register cleanup functions with toil that will be run in these situations. We will add a `docker_call` and a corresponding `docker_cleanup` function in toil-lib that can be used by any user.

**Comment by hannes-ucsc** [comment]
Try to design this mechanism in a way that let's us do arbitrary per-job cleanup: Let's say we have a job function `foo`:

```py
def foo(job):
    job.defer(myCleanUp) # register a function to be called when this function returns or raises
    job.recover(myRecovery) # register a function to be invoked before this job is to be retried
    …
```

The particulars would have to be discussed further. For example, we might want Toil to guarantee that either the deferred function or the clean-up function are going to be invoked, but never both. Not sure.

`docker_call()` would then have to be extended to exploit this mechanism. The deferred function would `docker rm` one specific container—the one that will be created by this invocation of `docker_call`, the recovery function would list all containers and kill the ones that could have been created by any number of docker_call invocations.
```

**Comment by arkal** [comment]
can we discuss this tomorrow after lunch?

**Comment by fnothaft** [comment]
@hannes-ucsc one thing to keep in mind is that the clean up function would need to run on the same node as the original job, right? This is hard to do on some batch systems, without "going around" the batch system (i.e., ssh'ing directly to the node).

**Comment by arkal** [comment]
yes which is why we need to chat. We were thinking of a generic cleanup that could be registered with toil. Then any job on a node can check for failed jobs and handle them

**Comment by hannes-ucsc** [comment]
If you find a crashed job (a job whose worker process crashed before being able to run the deferred functions), you can load its recovery function and run it. You are guaranteed to run it on the same node as the one the job was run on because otherwise you wouldn't have found the crashed job by looking at its left-over caching state. Note that this looking for crashed jobs is not happens whenever a job is run, not just on retries of the crashed job. IOW, as long as there are any jobs to run on a node, cleanup will happen.

The problem arises during the end game when only zombies are left for retry. The retries may not be run on the same node as a previous attempt so on some worker nodes there is nothing triggering the code that looks for crashed jobs. I feel like this is where BatchSystem.supportsWorkerCleanup() comes in. The recovery, or garbage collection, of crashed jobs happens one last time during worker cleanup, on every node.

And yes, discuss over lunch.

**Comment by fnothaft** [comment]
This generally sounds good. The one thing that I would make clear when this is actually implemented is that the cleanup function isn't guaranteed to be executed (that whole `BatchSystem.supportsWorkerCleanup()`) bit. Essentially, what we're providing isn't a _job_ cleanup function, it is a _job invocation_ cleanup function. In a job cleanup function, one might do something like try to clean up a failed upload to S3, or etc. A job invocation cleanup function would just clean up the state of the job left on the node where it ran. Catch my drift?

> And yes, discuss over lunch.

Obviously, I won't catch you all at lunch ;). Can you post a summary of whatever you discuss back here? TIA!

**Comment by arkal** [comment]
>  In a job cleanup function, one might do something like try to clean up a failed upload to S3, or etc. A job invocation cleanup function would just clean up the state of the job left on the node where it ran. Catch my drift?

Then we need to think about "how failed is a job". Either that or we write a robust cleanup function.

**Comment by fnothaft** [comment]
> Then we need to think about "how failed is a job". Either that or we write a robust cleanup function.

What exactly do you mean here? I didn't quite follow.

**Comment by arkal** [comment]
> clean up a failed upload to S3
you will clean up an upload  only if there was actually an upload that failed, right? Was there an upload in the job fn? what stage of the job did the job fail on? Did it reach an upload stage? If so, how do we find and handle the upload failure; If not what do we do?

Things like that.

**Comment by fnothaft** [comment]
> Things like that.

Oh yeah, I agree. That being said, you could imagine having a _job_ cleanup function that is guaranteed to run on _any_ host after the job fails (perhaps, guaranteed to run before the next retry of said failed job), and then having a _job invocation_ cleanup function that is guaranteed to run the next time a job starts on the host where said job invocation failed. The former would be sufficient for cleaning up things like failed uploads, while the latter would be sufficient for cleaning up things like ghost containers.

**Comment by arkal** [comment]
> you could imagine having a job cleanup function that is guaranteed to run on any host after the job fails

That is what we are targeting, i believe. The plan is to have a check at every job start that looks at the state of all jobs on the worker, and deals with dead/killed jobs and their containers(through the cleanup fn). This sort of ties up to 
>  a job invocation cleanup function that is guaranteed to run the next time a job starts on the host where said job invocation failed

as well

**Comment by fnothaft** [comment]
SGTM!

**Comment by hannes-ucsc** [comment]
Sounds like you guys don't think we need the distinction between deferred and recovery functions. I tend to agree.

**Comment by arkal** [comment]
Hannes and I had a talk today and came up with a great idea of pickling/dilling job specific cleanup functions into the cache state file in the form of a list such that the cleanup of a job is on the node and is not integrated into the job store. Jobs detected to have been failed can be handled easily since the cleanup function is already defined on the node.

**Comment by fnothaft** [comment]
+1, thanks for the update @arkal

**Comment by hannes-ucsc** [comment]
The commits should not have referred to this issue as (resolves ) but rather (connected to). We still need the fix to `docker_call` in `toil-lib` to ultimately resolve this.

**Comment by arkal** [comment]
Hmmm.... my bad!

**Comment by hannes-ucsc** [comment]
No worries. I think you can make cross-repo issue references in commits so the toil-lib commit can close a ticket in toil. The syntax is `(resolves #BD2KGenomics/Toil#987)` I believe.

**Comment by cket** [comment]
@arkal can this ticket be closed?

**Comment by arkal** [comment]
Not yet. This was partially resolved in #1128 but will be completed in #1189 .

**Comment by cket** [comment]
@arkal Can you close this if this is done now? #1189 was merged.

**Comment by arkal** [comment]
yeah this can be closed. Does @fnothaft have anything to add on?
